dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java

```diff
@@ -0,0 +1,182 @@
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Sets;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.ClassParser;
+import org.apache.bcel.classfile.Constant;
+import org.apache.bcel.classfile.ConstantMethodref;
+import org.apache.bcel.classfile.ConstantNameAndType;
+import org.apache.bcel.classfile.ConstantPool;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Method;
+
+/**
+ * This class reads a Java class file to analyze following attributes:
+ *
+ * <ol>
+ *   <li>source (defined methods) via Method fields in the class, and</li>
+ *   <li>targets (what's attempted to be invoked) via the constant pool table of the class</li>
+ * </ol>
+ */
+class ClassDumper {
+
+  private static List<FullyQualifiedMethodSignature> listMethodReferences(JavaClass javaClass) {
+    List<FullyQualifiedMethodSignature> methodReferences = new ArrayList<>();
+    ConstantPool constantPool = javaClass.getConstantPool();
+    Constant[] constants = constantPool.getConstantPool();
+    List<Constant> methodrefConstants = Arrays.stream(constants)
+            .filter(Predicates.notNull())
+            .filter(constant -> constant.getTag() == Const.CONSTANT_Methodref)
+            .collect(Collectors.toList());
+    for (Constant constant : methodrefConstants) {
+      ConstantMethodref constantMethodref = (ConstantMethodref) constant;
+      String className = constantMethodref.getClass(constantPool);
+      int nameAndTypeIndex = constantMethodref.getNameAndTypeIndex();
+      Constant constantAtNameAndTypeIndex = constantPool.getConstant(nameAndTypeIndex);
+      if (!(constantAtNameAndTypeIndex instanceof ConstantNameAndType)) {
+        // This constant_pool entry must be a CONSTANT_NameAndType_info
+        // as specified https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.4.2
+        throw new RuntimeException(
+            "Failed to lookup nameAndType constant indexed " + nameAndTypeIndex
+                + ". This class file is not compliant with CONSTANT_Methodref_info specification");
+      }
+      ConstantNameAndType constantNameAndType = (ConstantNameAndType) constantAtNameAndTypeIndex;
+      String methodName = constantNameAndType.getName(constantPool);
+      String descriptor = constantNameAndType.getSignature(constantPool);
+      FullyQualifiedMethodSignature methodref = new FullyQualifiedMethodSignature(className,
+          methodName, descriptor);
+      methodReferences.add(methodref);
+    }
+    return methodReferences;
+  }
+
+  /**
+   *  Lists all Methodref entries defined in a constant pool table of a class file
```

Comment:
This seems to be describing the method in terms of its implementation, but I think the first sentence should be focused on the the caller cares about. Something like: "Lists all method references from the class file." Then, after that you could talk about the implementation. In particular, I don't think we should talk about `Methodref` as a concept (that is part of BCEL and is an implementation detail for us). 

(Adjust the first sentences for the javadocs of the other methods in the same way)

Reply:
Nice. I'll follow it.

Reply:
Now these first sentences are simplified and without implementation detail:

>   *  Lists all method references from the class file. The output corresponds to
>  *  CONSTANT_Methodref_info entries its constant pool table